### ISSUE TITLE
aten.topk costs a fixed 1.75 ms on device when the rest of the graph is on the ANE

### ISSUE THREAD

**Issue by john-rocky** [apple/coreai-torch#66]
## Ask

Two things, and the first is the one this repo can act on.

1. Document where the ANE partition boundary falls, so authors can structure a graph around it. `docs/guides/` covers composite ops, custom Metal kernels and custom lowerings, but nothing says which ops keep a graph on the ANE.
2. A top-k primitive that survives the ANE partition. That is a dialect change rather than a lowering fix, so it may sit outside what this repo accepts — if it belongs in Feedback Assistant instead, say so and I will move it.

## What I measured

`aten.topk` costs a fixed 1.75 ms on an iPhone 17 Pro whenever the rest of the graph is on the ANE. On the GPU the same op costs 0.05 ms.

The model is YOLO26n from ultralytics 8.4.128, exported through coreai-torch 0.4.2 at 640×640, fp16. iPhone 17 Pro, iOS 27.0 beta 6 (24A5418b), AOT-compiled for h18p, input NDArray allocated once outside the timed loop, three interleaved blocks of 50 iterations, thermal state nominal, block-to-block spread under 1%.

Staging the graph one step at a time:

| graph | ANE-preferred | GPU-only |
|---|---|---|
| backbone + neck + DFL + max over classes | 1.23 ms | 3.16 ms |
| the same, plus one `topk` (axis length 8400) | 2.99 ms | 3.20 ms |
| delta from adding `topk` | **+1.75 ms** | **+0.05 ms** |

The GPU-only column is the control. The op is cheap; crossing to it is not.

## Why it is not the sort

Three measurements rule out the obvious readings.

`max_det` 300 → 10 changes the full model from 3.09 ms to 3.00 ms. Cutting k by 30× buys 3%, so k is not the cost.

A second `topk` over an axis of 24,000 adds 0.06 ms on top of the first. Tripling the axis length costs 3% of what the first one cost, so N is not the cost either.

`replace_topk` in `coreai_torch/_aten_to_core.py:2738` emits `coreai.sort` and `coreai.argsort` over the whole axis, then slices k off each. One `argsort` plus a `gather_along_axis` is equivalent and halves the sorting. I built that patch and measured it: 3.096 ms → 3.062 ms, 1.8%. I am not proposing it. It fixes something real and does not touch what is expensive.

Only a fixed cost charged at the first non-ANE op fits all four numbers.

## Caveat on the mechanism

The partition explanation is inference. It rests on the fixed-cost signature and the GPU-only control, not on a placement report — I could not find an API in the public Python runtime that says which ops ran on which unit. `ComputeUnitKind` is a request, and `Profiler` / `IntermediateLogger` / `LogEvent` did not expose placement in what I read. If that API exists I would rather confirm the mechanism than infer it, and knowing about it would also make the documentation ask above much smaller.

## Reproducer

```python
import torch, coreai_torch
from coreai_torch import TorchConverter
from ultralytics import YOLO
from ultralytics.nn.modules import Detect

model = YOLO("yolo26n.pt").model.float().eval()
for m in model.modules():
    if isinstance(m, Detect):
        m.export, m.format, m.dynamic, m.max_det = True, "coreai", False, 300
        m.agnostic_nms, m.xyxy, m.shape = False, False, None
model = model.fuse()

# STAGE = "body" keeps the graph on the ANE; STAGE = "topk" adds one topk.
STAGE = "topk"
for m in model.modules():
    if isinstance(m, Detect):
        def post(preds, nc=m.nc, k=m.max_det, stage=STAGE):
            _, scores = preds.split([4, nc], dim=-1)
            s = scores.max(dim=-1)[0]
            return s if stage == "body" else s.topk(k, dim=1)[0]
        m.postprocess = post

class Wrap(torch.nn.Module):
    def __init__(self, m): super().__init__(); self.m = m
    def forward(self, x):
        y = self.m(x)
        return y[0] if isinstance(y, (list, tuple)) else y

ex = torch.rand(1, 3, 640, 640).half()
w = Wrap(model).eval().half()
with torch.no_grad():
    ep = torch.export.export(w, (ex,)).run_decompositions(coreai_torch.get_decomp_table())

c = TorchConverter()
c.add_exported_program(ep, entrypoint_name="main", input_names=["image"], output_names=["out0"])
prog = c.to_coreai(); prog.optimize(); prog.save_asset(f"{STAGE}.aimodel")
```

Then compile each stage for both preferences and time them on device:

```
xcrun coreai-build compile topk.aimodel --platform iOS --preferred-compute neural-engine --architecture h18p --output ane/
xcrun coreai-build compile topk.aimodel --platform iOS --preferred-compute gpu           --architecture h18p --output gpu/
```

## Context

This came out of adding Core AI export to Ultralytics ([ultralytics/ultralytics#25620](https://github.com/ultralytics/ultralytics/issues/25620)). The detection body converts and runs at parity with Core ML on the same device — 1.32 ms against 1.32 ms with the postprocess out of the graph. With the end2end head in the graph it is 3.06 ms against 1.53 ms, and the entire gap is the row above.

Two smaller notes from the same work, in case they are useful. `coreai-torch` has no `aten.remainder.Scalar` lowering; decomposing it to `x - s * torch.div(x, s, rounding_mode="floor")` for integer dtypes matches torch exactly and is the only op that blocked this model. And an asset compiled with `--preferred-compute none` fails to specialize to CPU-only on device with `failedToSpecialize`, while the `gpu` and `neural-engine` bundles load fine — I can file that separately if it is not expected.
